README.md
bs58
====

[![build status](https://secure.travis-ci.org/cryptocoinjs/bs58.png)](http://travis-ci.org/cryptocoinjs/bs58)

JavaScript component to compute base 58 encoding. This encoding is typically used for crypto currencies such as Bitcoin.


Install
-------

    npm i --save bs58


API
---

### encode(input)

`input` must be a [Buffer](http://nodejs.org/api/buffer.html) or an `Array`. It returns a `string`.

**example**:

```js
var bs58 = require('bs58')

var unencodedData = "003c176e659bea0f29a3e9bf7880c112b1b31b4dc826268187"
var out = bs58.encode(new Buffer(unencodedData, 'hex'))
console.log(out) // => 16UjcYNBG9GTK4uq2f7yYEbuifqCzoLMGS
```


### decode(input)

`input` must be a base 58 encoded string. Returns an `Array`.

**example**:

```js
var bs58 = require('bs58')

var address = "16UjcYNBG9GTK4uq2f7yYEbuifqCzoLMGS"
var out = bs58.decode(address)
console.log(out.toString())
// => 0,60,23,110,101,155,234,15,41,163,233,191,120,128,193,18,177,179,27,77,200,38,38,129,135

// if using Node.js or browserify
console.log(new Buffer(out).toString('hex'))
// => 003c176e659bea0f29a3e9bf7880c112b1b31b4dc826268187
```

Hack / Test
-----------

Uses JavaScript standard style. Read more:

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)


Credits
-------
- [Mike Hearn](https://github.com/mikehearn) for original Java implementation
- [Stefan Thomas](https://github.com/justmoon) for porting to JavaScript
- [Stephan Pair](https://github.com/gasteve) for buffer improvements
- [Daniel Cousens](https://github.com/dcousens) for cleanup and merging improvements from bitcoinjs-lib
- [Jared Deckard](https://github.com/deckar01) for killing `bigi` as a dependency


License
-------

MIT


CHANGELOG.md
3.0.0 / 2015-08-18
------------------
- refactored module into generic [`base-x`](https://github.com/cryptocoinjs/base-x).

2.0.1 / 2014-12-23
------------------
- performance boost in `encode()` [#10](https://github.com/cryptocoinjs/bs58/pull/10)

2.0.0 / 2014-10-03
------------------
- `decode` now returns an `Array` instead of `Buffer` to keep things simple. [Daniel Cousens / #9](https://github.com/cryptocoinjs/bs58/pull/9)

1.2.1 / 2014-07-24
------------------
* speed optimizations [Daniel Cousens / #8](https://github.com/cryptocoinjs/bs58/pull/8)

1.2.0 / 2014-06-29
------------------
* removed `bigi` dep, implemented direct byte conversion [Jared Deckard / #6](https://github.com/cryptocoinjs/bs58/pull/6)

1.1.0 / 2014-06-26
------------------
* user `Buffer` internally for calculations, providing cleaner code and a performance increase. [Daniel Cousens](https://github.com/cryptocoinjs/bs58/commit/129c71de8bc1e36f113bce06da0616066f41c5ca)

1.0.0 / 2014-05-27
------------------
* removed `binstring` dep, `Buffer` now only input to `encode()` and output of `decode()`
* update `bigi` from `~0.3.0` to `^1.1.0`
* added travis-ci support
* added coveralls support
* modified tests and library to handle fixture style testing (thanks to bitcoinjs-lib devs and [Daniel Cousens](https://github.com/dcousens))


0.3.0 / 2014-02-24
------------------
* duck type input to `encode` and change output of `decode` to `Buffer`.


0.2.1 / 2014-02-24
------------------
* removed bower and component support. Closes #1
* convert from 4 spaces to 2


0.2.0 / 2013-12-07
------------------
* renamed from `cryptocoin-base58` to `bs58`


0.1.0 / 2013-11-20
------------------
* removed AMD support


0.0.1 / 2013-11-04
------------------
* initial release


test/bs58.js
/* global describe, it */

var assert = require('assert')
var base58 = require('../')

var fixtures = require('./fixtures.json')

describe('base58', function () {
  describe('encode', function () {
    fixtures.valid.forEach(function (f) {
      it('can encode ' + f.hex, function () {
        var actual = base58.encode(new Buffer(f.hex, 'hex'))

        assert.strictEqual(actual, f.string)
      })
    })
  })

  describe('decode', function () {
    fixtures.valid.forEach(function (f) {
      it('can decode ' + f.string, function () {
        var actual = new Buffer(base58.decode(f.string)).toString('hex')

        assert.strictEqual(actual, f.hex)
      })
    })

    fixtures.invalid.forEach(function (f) {
      it('throws on ' + f.description, function () {
        assert.throws(function () {
          base58.decode(f.string)
        }, /Non-base58 character/)
      })
    })
  })

  describe('valid', function () {
    fixtures.valid.forEach(function (f) {
      it('returns true for ' + f.string, function () {
        assert(base58.valid(f.string))
      })
    })

    fixtures.invalid.forEach(function (f) {
      it('returns false for ' + f.description + ' (' + f.string + ')', function () {
        assert(!base58.valid(f.string))
      })
    })
  })
})


test/B58.test.js
var base58 = require('../lib/B58')
  , conv = require('convert-hex')

require('terst')

describe('+ encode(input)', function() {
  describe('> when input a byte array', function() {
    it('should convert to the base58 encoding', function() {
      var privateKeyAndChecksum = "801184CD2CDD640CA42CFC3A091C51D549B2F016D454B2774019C2B2D2E08529FD206EC97E"
      var privateKeyWIF = base58.encode(conv.hexToBytes(privateKeyAndChecksum))
      EQ (privateKeyWIF, "5Hx15HFGyep2CfPxsJKe2fXJsCVn5DEiyoeGGF6JZjGbTRnqfiD")

      var unencodedAddress = "003c176e659bea0f29a3e9bf7880c112b1b31b4dc826268187"
      var address = base58.encode(conv.hexToBytes(unencodedAddress))
      EQ (address, "16UjcYNBG9GTK4uq2f7yYEbuifqCzoLMGS")
    })
  })
})

describe('+ decode(input)', function() {
  describe('> when a byte array is input', function() {
    it('should convert to a a byte array', function() {
       var address = "16UjcYNBG9GTK4uq2f7yYEbuifqCzoLMGS"
    EQ (conv.bytesToHex(base58.decode(address)), "003c176e659bea0f29a3e9bf7880c112b1b31b4dc826268187")
    })
  })
})




lib/B58.js
!function(globals) {
'use strict'

var BigInteger = null
var base58 = {} //exports

if (typeof module !== 'undefined' && module.exports) { //CommonJS
  try { //Node.js
    BigInteger = require('cryptocoin-bigint')
  } catch (e) { //Component
    BigInteger = require('bigint')
  }
  module.exports = base58
} else {
  BigInteger = globals.BigInteger
  globals.base58 = base58
}


var alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
var base = BigInteger.valueOf(58);

var positions = {};
for (var i=0 ; i < alphabet.length ; ++i) {
    positions[alphabet[i]] = i;
}

// Convert a byte array to a base58-encoded string.
// Written by Mike Hearn for BitcoinJ.
//   Copyright (c) 2011 Google Inc.
// Ported to JavaScript by Stefan Thomas.
base58.encode = function (input) {
    var bi = BigInteger.fromByteArrayUnsigned(input);
    var chars = [];

    while (bi.compareTo(base) >= 0) {
        var mod = bi.mod(base);
        chars.push(alphabet[mod.intValue()]);
        bi = bi.subtract(mod).divide(base);
    }
    chars.push(alphabet[bi.intValue()]);

    // Convert leading zeros too.
    for (var i = 0; i < input.length; i++) {
        if (input[i] == 0x00) {
            chars.push(alphabet[0]);
        } else break;
    }

    return chars.reverse().join('');
}

// decode a base58 string into a byte array
// input should be a base58 encoded string
// @return Array
base58.decode = function (input) {

  var base = BigInteger.valueOf(58);

  var length = input.length;
  var num = BigInteger.valueOf(0);
  var leading_zero = 0;
  var seen_other = false;
  for (var i=0; i<length ; ++i) {
      var char = input[i];
      var p = positions[char];

      // if we encounter an invalid character, decoding fails
      if (p === undefined) {
          throw new Error('invalid base58 string: ' + input);
      }

      num = num.multiply(base).add(BigInteger.valueOf(p));

      if (char == '1' && !seen_other) {
          ++leading_zero;
      }
      else {
          seen_other = true;
      }
  }

  var bytes = num.toByteArrayUnsigned();

  // remove leading zeros
  while (leading_zero-- > 0) {
      bytes.unshift(0);
  }

  return bytes;
}


}(this);

lib/bs58.js
// Base58 encoding/decoding
// Originally written by Mike Hearn for BitcoinJ
// Copyright (c) 2011 Google Inc
// Ported to JavaScript by Stefan Thomas
// Merged Buffer refactorings from base58-native by Stephen Pair
// Copyright (c) 2013 BitPay Inc

var ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
var ALPHABET_MAP = {}
for (var i = 0; i < ALPHABET.length; ++i) {
  ALPHABET_MAP[ALPHABET.charAt(i)] = i
}
var BASE = ALPHABET.length

function encode (buffer) {
  if (buffer.length === 0) return ''

  var i, j
  var digits = [0]

  for (i = 0; i < buffer.length; ++i) {
    for (j = 0; j < digits.length; ++j) digits[j] <<= 8

    digits[0] += buffer[i]

    var carry = 0
    for (j = 0; j < digits.length; ++j) {
      digits[j] += carry

      carry = (digits[j] / BASE) | 0
      digits[j] %= BASE
    }

    while (carry) {
      digits.push(carry % BASE)

      carry = (carry / BASE) | 0
    }
  }

  // deal with leading zeros
  for (i = 0; buffer[i] === 0 && i < buffer.length - 1; ++i) digits.push(0)

  // convert digits to a string
  var str = ''
  for (i = digits.length - 1; i >= 0; --i) {
    str += ALPHABET[digits[i]]
  }

  return str
}

function decode (string) {
  if (string.length === 0) return []

  var i, j
  var bytes = [0]

  for (i = 0; i < string.length; ++i) {
    var c = string[i]
    if (!(c in ALPHABET_MAP)) throw new Error('Non-base58 character')

    for (j = 0; j < bytes.length; ++j) bytes[j] *= BASE
    bytes[0] += ALPHABET_MAP[c]

    var carry = 0
    for (j = 0; j < bytes.length; ++j) {
      bytes[j] += carry

      carry = bytes[j] >> 8
      bytes[j] &= 0xff
    }

    while (carry) {
      bytes.push(carry & 0xff)

      carry >>= 8
    }
  }

  // deal with leading zeros
  for (i = 0; string[i] === '1' && i < string.length - 1; ++i) bytes.push(0)

  return bytes.reverse()
}

function valid (value) {
  return /^[1-9A-HJ-NP-Za-km-z]*$/.test(value)
}

module.exports = {
  encode: encode,
  decode: decode,
  valid: valid
}


lib/base58.js
!function(globals) {
'use strict'

var BigInteger = null
var base58 = {} //exports

//*** UMD BEGIN
if (typeof define !== 'undefined' && define.amd) { //require.js / AMD
  define(['cryptocoin-bigint'], function(_BigInteger) {
    BigInteger = _BigInteger
    return base58
  })
} else if (typeof module !== 'undefined' && module.exports) { //CommonJS
  try { //Node.js
    BigInteger = require('cryptocoin-bigint')
  } catch (e) { //Component
    BigInteger = require('bigint')
  }
  module.exports = base58
} else {
  BigInteger = globals.BigInteger
  globals.base58 = base58
}
//*** UMD END

var alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
var base = BigInteger.valueOf(58);

var positions = {};
for (var i=0 ; i < alphabet.length ; ++i) {
    positions[alphabet[i]] = i;
}

// Convert a byte array to a base58-encoded string.
// Written by Mike Hearn for BitcoinJ.
//   Copyright (c) 2011 Google Inc.
// Ported to JavaScript by Stefan Thomas.
base58.encode = function (input) {
    var bi = BigInteger.fromByteArrayUnsigned(input);
    var chars = [];

    while (bi.compareTo(base) >= 0) {
        var mod = bi.mod(base);
        chars.push(alphabet[mod.intValue()]);
        bi = bi.subtract(mod).divide(base);
    }
    chars.push(alphabet[bi.intValue()]);

    // Convert leading zeros too.
    for (var i = 0; i < input.length; i++) {
        if (input[i] == 0x00) {
            chars.push(alphabet[0]);
        } else break;
    }

    return chars.reverse().join('');
}

// decode a base58 string into a byte array
// input should be a base58 encoded string
// @return Array
base58.decode = function (input) {

  var base = BigInteger.valueOf(58);

  var length = input.length;
  var num = BigInteger.valueOf(0);
  var leading_zero = 0;
  var seen_other = false;
  for (var i=0; i<length ; ++i) {
      var char = input[i];
      var p = positions[char];

      // if we encounter an invalid character, decoding fails
      if (p === undefined) {
          throw new Error('invalid base58 string: ' + input);
      }

      num = num.multiply(base).add(BigInteger.valueOf(p));

      if (char == '1' && !seen_other) {
          ++leading_zero;
      }
      else {
          seen_other = true;
      }
  }

  var bytes = num.toByteArrayUnsigned();

  // remove leading zeros
  while (leading_zero-- > 0) {
      bytes.unshift(0);
  }

  return bytes;
}


}(this);